### PR TITLE
Fix/1468 add reset for forms

### PR DESCRIPTION
**Pull request by Diaan**

### CODE REVIEW

File: packages/components/form/src/form.stories.ts [hunk 1]
```diff
@@ -27,7 +28,7 @@ export default {
     reportValidity: false
   },
   render: args => {
-    const { buttons, disabled, fields, reportValidity, value } = args;
+    const { buttons, disabled, fields, reportValidity, value, reset } = args;
```
Comment (jpzwarte) [inline packages/components/form/src/form.stories.ts:31]: nitpicking, but please keep it sorted

File: packages/components/form/src/form.stories.ts [hunk 2]
```diff
@@ -41,6 +42,12 @@ export default {
       console.log(form?.reportValidity(), form?.value);
     };
 
+    const onReset = (): void => {
+      const form = document.querySelector('sl-form');
```
Comment (jpzwarte) [inline packages/components/form/src/form.stories.ts:46]: Can be a 1 liner

File: packages/components/form/src/form.stories.ts [hunk 3]
```diff
@@ -83,6 +91,23 @@ export const Basic: Story = {
...
+    reset: true,
+    reportValidity: true,
+    fields: () =>
+      html` <sl-form-field hint="Has value on load" label="Text field">
+          <sl-text-field name="input" placeholder="Placeholder" required value="Value set initially"></sl-text-field>
+        </sl-form-field>
+        <sl-form-field hint="Has no value on load" label="Text field">
+          <sl-text-field name="input2" placeholder="Placeholder" required></sl-text-field>
+        </sl-form-field>`
+    // value: {
```
Comment (jpzwarte) [inline packages/components/form/src/form.stories.ts:105]: Remove comment plz

File: packages/components/form/src/form.ts [hunk 4]
```diff
@@ -41,6 +41,8 @@ export class Form<T extends Record<string, unknown> = Record<string, unknown>> e
    */
   #value: T | undefined;
 
+  #initialValue: T | undefined;
```
Comment (jpzwarte) [inline packages/components/form/src/form.ts:44]: Please add a comment (even though the name is perfectly clear, we always have comments for variables)

File: packages/components/form/src/form.ts [hunk 5]
```diff
@@ -87,13 +89,16 @@ export class Form<T extends Record<string, unknown> = Record<string, unknown>> e
...
   get value(): T {
-    return this.controls.reduce((value, control) => {
+    const value = this.controls.reduce((value, control) => {
       if (control.name) {
         setValueByPath(value, control.name, control.formValue);
       }
-
       return value;
     }, {}) as T;
+    if (!this.#initialValue && !!value) {
```
Comment (jpzwarte) [inline packages/components/form/src/form.ts:98]: Please add a newline before the if statement

File: packages/components/form/src/form.ts [hunk 6]
```diff
@@ -87,13 +89,16 @@ export class Form<T extends Record<string, unknown> = Record<string, unknown>> e
...
+      this.#initialValue = value;
+    }
+    return value;
```
Comment (jpzwarte) [inline packages/components/form/src/form.ts:101]: Add a newline before the return statement

File: packages/components/form/src/form.ts [hunk 7]
```diff
@@ -107,6 +112,11 @@ export class Form<T extends Record<string, unknown> = Record<string, unknown>> e
     }
   }
 
+  override firstUpdated(changes: PropertyValues<this>): void {
+    super.firstUpdated(changes);
+    this.#initialValue = this.#value;
```
Comment (jpzwarte) [inline packages/components/form/src/form.ts:117]: Should we have 2 places where we set the `#initialValue`?

**Review by jpzwarte — APPROVED** [review]

File: packages/components/form/src/form.spec.ts [hunk 8]
```diff
@@ -175,6 +175,43 @@ describe('sl-form', () => {
...
+        </sl-form>
+      `);
+      // Give the form time to register the controls
+      await new Promise(resolve => setTimeout(resolve, 600));
+    });
+
+    it('should reset the form', async () => {
+      expect(el.value).to.deep.equal({ foo: 'lorem', bar: '' });
+      el.querySelector<HTMLElement>('sl-text-field[name="foo"]')?.focus();
+      await sendKeys({ press: 'Backspace' });
```
[the remark this answers is not in the document]
Comment (Diaan) [reply packages/components/form/src/form.spec.ts:198]: no, that didn't work, i tried it with one backspace at first

File: packages/components/form/src/form.ts [hunk 9]
```diff
@@ -126,6 +136,19 @@ export class Form<T extends Record<string, unknown> = Record<string, unknown>> e
     return this.controls.map(c => c.reportValidity()).every(Boolean);
   }
 
+  reset(): void {
```
Comment (anna-lach) [inline packages/components/form/src/form.ts:139]: Can you please add the description that will be shown on the documentation website as well?

image

File: packages/components/form/src/examples.stories.ts [hunk 10]
```diff
@@ -25,11 +25,17 @@ export default {
 export const LogIn: Story = {
   render: () => {
     const onSubmit = (event: Event & { target: HTMLElement }): void => {
+      event.preventDefault();
```
[the remark this answers is not in the document]
Comment (Diaan) [reply packages/components/form/src/examples.stories.ts:28]: I wanted to try to get it working with button `type="submit"` (and i didn't want it to accually submit the form) but that didn't work...

File: packages/components/form/src/form.ts (hunk 7 above)
Comment (Diaan) [reply packages/components/form/src/form.ts:117]: one is for when the value is set, the other for when it isn't set. No idea why, but it doesn't work with just one

**Pull request merged**